llvm/lib/SYCLLowerIR/SYCLPropagateJointMatrixUsage.cpp

```diff
@@ -0,0 +1,244 @@
+//===---- SYCLPropagateJointMatrixUsage.cpp - SYCLPropagateJointMatrixUsage Pass
+//--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass propagates optional kernel features metadata through a module call graph
+// for sycl_ext_oneapi_matrix extension
+//
+// The pass consists of three main steps:
+//
+// I. It builds Function -> string of joint matrix types and sizes values
+// mapping for usage in step II
+// II. Propagates all the values from step I. to the top of the call graph
+// III. Generates metadata with values of joint matrix types and sizes
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLPropagateJointMatrixUsage.h"
+
+#include "llvm/ADT/SmallString.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/IntrinsicInst.h"
+
+#include <set>
+
+using namespace llvm;
+
+namespace {
+
+/// Returns true if the function is a SPIRV or SYCL builtin, e.g.
+/// _Z28__spirv_GlobalInvocationId_xv
+/// NB! This function was copied from sycl-post-link/ModuleSplitter.cpp and the
+/// definition of entry point (i.e. implementation of the function) should be in
+/// sync between those two.
+bool isSpirvSyclBuiltin(StringRef FName) {
+  if (!FName.consume_front("_Z"))
+    return false;
+  // now skip the digits
+  FName = FName.drop_while([](char C) { return std::isdigit(C); });
+
+  return FName.startswith("__spirv_") || FName.startswith("__sycl_");
+}
+
+bool isEntryPoint(const Function &F) {
+  // Skip declarations, we can't analyze them
+  if (F.isDeclaration()) {
+    // F.print(outs());
+    return false;
+  }
+
+  // Kernels are always considered to be entry points
+  if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+    return true;
+
+  // FIXME: sycl-post-link allows to disable treating SYCL_EXTERNAL's as entry
+  // points - do we need similar flag here?
```

Comment:
I don't think so - since it's `sycl-post-link`'s choice of whether or these as entry points or not, then even if this information is propagated, it can be discarded by sycl-post-link. By the way, can a test be added to ensure the joint matrix features are propagated?

Reply:
Sounds reasonable.

> By the way, can a test be added to ensure the joint matrix features are propagated to SYCL_EXTERNAL functions?

The test already covers SYCL_EXTERNAL functions.

Reply:
Oh, I actually didn't see that some of the functions in `call-graph-joint-matrix.ll` had `sycl-module-id`. Even though, the functions that do have `sycl-module-id` would have had joint matrix attributes propagated to them anyways because they are called from `spir_kernel` functions. The case that I had in mind and that was not covered in `call-graph-joint-matrix.ll` is when a `spir_func` with `sycl-module-id` that is not called from a `spir_kernel` has joint matrix features.

Reply:
Thanks, added such case